core/src/main/scala/dagr/core/execsystem/TaskManagerLike.scala

```diff
@@ -0,0 +1,177 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package dagr.core.execsystem
+
+import dagr.core.execsystem.TaskManagerLike.BaseInfoAndNode
+import dagr.core.tasksystem.Task
+import dagr.core.util.BiMap
+
+object TaskManagerLike {
+  abstract class BaseInfoAndNode(info: TaskExecutionInfo, node: BaseGraphNode)
+}
+
+/** A generic template for task managers */
+trait TaskManagerLike {
+
+  /** Gets the task associated with the identifier, if any
+    *
+    * @param id the task id
+    * @return the task associate with the id if found, None otherwise
+    */
+  def taskFor(id: BigInt): Option[Task]
+
+  /** Get the task's execution information if it exists.
+    *
+    * @param id the task identifier.
+    * @return the task execution information if the task is managed, None otherwise
+    */
+  def taskExecutionInfoFor(id: BigInt): Option[TaskExecutionInfo]
+
+  /** Get the task's execution information if it exists.
+    *
+    * @param task the task.
+    * @return the task execution information if the task is managed, None otherwise
+    */
+  def taskExecutionInfoFor(task: Task): Option[TaskExecutionInfo]
+
+  /** Get the task's associated [[TaskStatus]].
+    *
+    * @param task the task.
+    * @return the task status if the task is managed, None otherwise.
+    */
+  def taskStatusFor(task: Task): Option[TaskStatus.Value]
+
+
+  /** Get the task's associated [[TaskStatus]].
+    *
+    * @param id the task identifier.
+    * @return the task status if the task is managed, None otherwise.
+    */
+  def taskStatusFor(id: BigInt): Option[TaskStatus.Value]
+
+
+  /** Get the task identifier for the given task.
+    *
+    * @param task the task.
+    * @return the task identifier, None if the task is not being managed.
+    */
+  def taskIdFor(task: Task): Option[BigInt] = task.id
+
+  /** Get the task identifiers for all tracked tasks.
+    *
+    * @return the task ids
+    */
+  def taskIdsFor(): Iterable[BigInt]
+
+  /** Gets the task for the given id.  If there is no task being tracked
+    * with that id, throws a [[NoSuchElementException]].
+    *
+    * @param id the id to lookup.
+    * @return the associated task.
+    */
+  def apply(id: BigInt): Task
+
+  /** Gets the task execution info for the given task.  If the task is not
+    *  being tracked, throws a [[NoSuchElementException]].
+    *
+    * @param task the task to lookup.
+    * @return the associated task execution info.
+    */
+  def apply(task: Task): BaseInfoAndNode
```

Comment:
I don't love having these two different map-like `apply()` methods in the same class.  When reading through TaskManager and finding things like `this(this(id)).info` I had to stop and puzzle out what it was doing.  Could we keep the `apply(id): Task` method, but rename this to `infoFor(task): BaseInfoAndNode`?